Add a guide to creating a native Windows application in Python (#1778)
pypa/packaging.python.org: source/guides/windows-applications-embedding.rst

integrates much better with the operating system if the application is delivered
as a native Windows executable. However, Python is not a natively compiled
language, and so does not create executables by default.

One way around this issue is to make your Python code into a library, and
declare one or more :ref:`script entry points <script-entry-points>` for the
library. When the library is installed, the installer will provide an executable
which invokes the Python interpreter, calling your entry point function. This is
a very effective solution, and is used by many Python applications. It is
supported by utilities such as ``pipx`` and ``uv tool``, which make managing
such entry points (and the virtual environments needed to support them) easy.

There are, however, some downsides to this approach. The entry point wrapper
results in a chain of processes being created - the wrapper itself, the virtual
environment redirector, and finally the Python interpreter. Creating all these
processes isn't cheap, and particularly for a small command line utility, it
impacts application startup time noticeably. Furthermore, the entry point

> wrapper is a standard executable with a zipfile attached - because of this, the

Reviewer: ```suggestion
wrapper is a standard executable with a ZIP file attached - because of this, the
```

users need to care about.

If any of these issues matter to you, you should consider writing your
application in Python, but then embedding it, and the Python interpreter, into a
native executable application that you can ship to your users. This does require
you to write a small amount of native code (typically in C), but the code is
mostly boilerplate and easy to maintain.

What will your application look like?
-------------------------------------

When embedding Python, it is not possible to create a "single file" application.
The Python interpreter itself is made up of multiple files, so you will need
those at a minimum. However, once you have decided to ship your application as
multiple files, it becomes very easy to structure your code.

There are basically three "parts" to an embedded application:

1. The main executable that the user runs.
2. The Python interpreter.
3. Your application code, written in Python.

You can, if you wish, dump all of those items into a single directory. However,
it is much easier to manage the application if you keep them separate.
Therefore, the recommended layout is::

>     Application directory
>         MyAwesomePythonApp.exe
>         interp
>             (embedded Python interpreter)
>         lib
>             (Python code implementing the application)

Reviewer: ```suggestion
    Application directory/
    ├── MyAwesomePythonApp.exe
    ├── interp/
    │   ├── (embedded Python interpreter)
    │   └── ...
    └── lib/
        ├── (Python code implementing the application)
        └── ...
```

installed correctly.

If that works, congratulations! You have successfully created the Python part of
your application.

The embedded interpreter
------------------------

You can download embeddable builds of Python from
https://www.python.org/downloads/windows/. You want the "Windows embeddable
package". There are usually 3 versions, for 64-bit, 32-bit and ARM64
architectures. Generally, you should use the 64-bit version unless you have a
specific need for one of the others (in which case, you will need to modify how
you compile the main application executable slightly, to match).

Simply unpack the downloaded zip file into the "interp" subdirectory of your
application layout.

In order for your embedded interpreter to be able to find your application code,
you should modify the ``python*._pth`` directory contained in the distribution. By
default it looks like this::

    python313.zip
    .

    # Uncomment to run site.main() automatically
    #import site

You need to add a single line, ``../lib``, after the line with the dot. The
resulting file will look like this::

    python313.zip
    .
    ../lib

    # Uncomment to run site.main() automatically
    #import site

> If you have put your application Python code somewhere else, this is the only

Reviewer: ```suggestion
If you have placed your application's Python code somewhere else, this is the only
```

thing you need to change. The file contains a list of directories (relative to
the interpreter directory) which will be added to Python's ``sys.path`` when
starting the interpreter.

The driver application
----------------------

This is the only part of your application that has to be written in C. The
application code should look like the following::

>     /* Include the Python headers */

Reviewer: Maybe use a code block fence with back ticks to allow syntax coloring.

````
```c
/* Include the Python headers */
#include <Python.h>
```
````

Author: Thanks, I wasn't aware that was possible.

Author: Looks like it isn't, that's Markdown. Maybe the code-block directive will work?

Author: OK, I think I finally got the syntax right.

Therefore, the recommended layout is::

    Application directory/
    ├── MyAwesomePythonApp.exe
    ├── interp/
    │   ├── (embedded Python interpreter)
    │   └── ...
    └── lib/
        ├── (Python code implementing the application)
        └── ...

The remainder of this guide will assume this layout.

How to build your application
=============================

Writing the Python code
-----------------------

Your Python application should be runnable by invoking a single function in your
application code. Typically, this function will be called ``main`` and will be
located in the root package of your application, but that isn't a hard
requirement. If you prefer to locate the function somewhere else, all you will
need to do is make a small modification to the wrapper code.

Your code can use 3rd party dependencies freely. These will be installed along
with your application.

You can create your application code in whatever way you prefer - there is no
particular need to make it buildable as a wheel. For the purposes of this
discussion, we will assume one of two project layouts:

1. A project which can be installed via ``pip install``, with all of its

>    dependencies listed in the ``pyproject.toml`` file, as normal.

Reviewer 2: ```suggestion
   dependencies listed in the :file:`pyproject.toml` file, as normal.
```

2. A Python application stored in a local directory. In this case, third party

>    dependencies are listed in a ``requirements.txt`` file.

Reviewer 2: ```suggestion
   dependencies are listed in a :file:`requirements.txt` file.
```

When you are ready to build your application, you need to copy all of the

> required code and dependencies into the ``lib`` directory of the application.

Reviewer 2: ```suggestion
required code and dependencies into the :file:`lib` directory of the application.
```

such modifications are out of scope for this guide. If you want to make such
changes, you should be familiar with the relevant parts of the Python C API
documentation and the Windows API.

Building the driver application
-------------------------------

To build the driver application, you will need a copy of Visual Studio, and a
full installation of the same version of Python as you are using for the
embedded interpreter. The reason for the full Python installation is that the
embedded version does not include the necessary C headers and library files to
build code using the Python C API.

It may be possible to use a C compiler other than Visual Studio, but if you wish
to do this, you will need to work out how to do the build, including the
necessary delay loading, yourself.

To compile the code, you need to know the location of the Python headers and
library files. You can get these locations from the interpreter as follows::

    import sysconfig

    print("Include files:", sysconfig.get_path("include"))
    print("Library files:", sysconfig.get_config_var("LIBDIR"))

To build your application, you can then simply use the following commands::

>     cl /c /Fo:main.obj main.c /I<Include File Location>

Author: Awesome, thanks! I've pushed an update, if that fixes this, then hopefully we're ready to merge!

you may have a lot of common Python code, with many different entry points.

This is fine - it's easy to modify the layout to cover these cases. You can have
as many executable files in the application directory as you want.  These can
all call their own entry point - they can even use completely independent
libraries of Python code, although in that case you'd need to add some code to
manipulate ``sys.path``.

The point is that the basic structure can be as flexible as you want it to be -
but it's better to start simple and add features as you need them, so that you
don't have to maintain code that handles cases you don't care about.

Potential Issues
================

Using tkinter
-------------

The embedded Python distribution does not include tkinter. If your application
needs a GUI, the simplest option is likely to be to use one of the other GUI
frameworks available from PyPI, such as PyQt or wxPython.

If your only option is tkinter, you will need to add a copy to the embedded
distribution. You can get the relevant files from a full Python installation
(which must be the same version as you're using for your application). The
files you need are:

> * ``_tkinter.pyd``, ``tcl*.dll``, ``tk*.dll`` and ``zlib1.dll`` from
>   ``{sys.prefix}\DLLs``
> * The ``tcl`` directory from ``{sys.prefix}``
> * The ``tkinter`` directory from ``{sys.prefix}\Lib``

Reviewer 2: ```suggestion
* :file:`_tkinter.pyd`, :file:`tcl*.dll`, :file:`tk*.dll` and :file:`zlib1.dll` from
  :file:`{sys.prefix}\DLLs`
* The :file:`tcl` directory from :file:`{sys.prefix}`
* The :file:`tkinter` directory from :file:`{sys.prefix}\Lib`
```

You should copy all of these into your embedded interpreter directory.

Subprocesses and ``sys.executable``
-----------------------------------

A common pattern in Python code is to run a Python subprocess using
``subprocess.run([sys.executable, ...])``. This will not work for an embedded
application, as ``sys.executable`` is your application, not the Python
interpreter.

The embedded distribution does contain a Python interpreter, which can be used
in cases like this, but you will need to locate it yourself::

    python_executable = Path(sys.executable).parent / ("interp/python.exe")

> If you are using the ``multiprocessing`` module, it has a specific method you

Reviewer 2: ```suggestion
If you are using the :py:mod:`multiprocessing` module, it has a specific method you
```

must use to configure it to work correctly in an embedded environment,

> documented `in the Library reference
> <https://docs.python.org/3.13/library/multiprocessing.html#multiprocessing.set_executable>`_.

Reviewer 2: I think this should work
```suggestion
documented :py:func:`in the Library reference <multiprocessing.set_executable>`.
```
